Secure Data Path test setup (SWG-208/209)

etienne-lms opened the pull request:
Adding a ION heap type dedicated to "secure data path buffer pools": buffers unmaped in linux world.

Add ION "umaped heap" creation support in ION dummy driver: possibly get "unmaped" heaps from DT.

This is currently a patch in the ION staging directory. ION implementation does not allow registering heap types from external (non ION) drivers.

This patchset was tested on k4.6 and k4.9 (ION integration changes)

SDP test setup on vexpress-qemu: see also https://github.com/OP-TEE/build/pull/112 and https://github.com/linaro-swg/bios_qemu_tz_arm/pull/10 and https://github.com/OP-TEE/optee_test/pull/148.

jockebech commented on drivers/staging/android/ion/Kconfig:
```diff
@@ -52,3 +52,39 @@ config ION_OF
 	  If using Ion and devicetree, you should say Y here
+
+config ION_UNMAPED
+	bool "Driver for an Ion heap type using unmapped heap"
```
s/Ion/ION/

etienne-lms replied:
ok for all....

jockebech commented on drivers/staging/android/ion/Kconfig:
```diff
@@ -52,3 +52,39 @@ config ION_OF
+	bool "Driver for an Ion heap type using unmapped heap"
+	depends on ION
+	help
+	  Provides a ION heap type for handling memory pool that are
```
s/a/an/

jockebech commented on drivers/staging/android/ion/Kconfig:
```diff
@@ -52,3 +52,39 @@ config ION_OF
+	depends on ION
+	help
+	  Provides a ION heap type for handling memory pool that are
+	  default not mapped to the linux kernel. This can be used
```
Here and several more places: `s/linux/Linux/`

etienne-lms replied:
ok

jockebech commented on drivers/staging/android/ion/Kconfig:
```diff
@@ -52,3 +52,39 @@ config ION_OF
+	  when linux executes in a non secure state and the target
+	  memory references are expected to be accesed from secure
+	  agent only.
+	  Ion "unmaped" heap rely on Ion UAPI to identify itself.
```
s/unmaped/unmapped/
s/Ion/ION/

jockebech commented on drivers/staging/android/ion/Kconfig:
```diff
@@ -52,3 +52,39 @@ config ION_OF
+	  memory references are expected to be accesed from secure
+	  agent only.
+	  Ion "unmaped" heap rely on Ion UAPI to identify itself.
+	  ION_HEAP_TYPE_UNMAPED = ION_HEAP_TYPE_CUSTOM + 1
```
s/ION_HEAP_TYPE_UNMAPED/ION_HEAP_TYPE_UNMAP**P**ED/

Also, isn't this a bit hackish? Wouldn't we want to get the new type included in the `enum ion_heap_type` (`drivers/staging/android/uapi/ion.h`) from start?

etienne-lms replied:
I intentionally did not change the ION UAPI, since it is UAPI. I preferred to insure "UNMAP**P**ED" heap type ID rely on already defined UAPI IDs, hence  CUSTOM + _something_, both in kernel and user lands.

jockebech commented on drivers/staging/android/ion/Kconfig:
```diff
@@ -52,3 +52,39 @@ config ION_OF
+	  If ION_DUMMY is enable, it can create a UNMAPED heap from
+	  DT nodes or kernel configuration.
+
+config ION_DUMMY_UNMAPED_HEAP
```
s/ION_DUMMY_UNMAPED_HEAP/ION_DUMMY_UNMAP**P**ED_HEAP/

jockebech commented on drivers/staging/android/ion/Kconfig:
```diff
@@ -52,3 +52,39 @@ config ION_OF
+	  DT nodes or kernel configuration.
+
+config ION_DUMMY_UNMAPED_HEAP
+	bool "Ion dummy driver define a Ion unmaped heap"
```
s/Ion/ION/ (two places on the same line)

jockebech commented on drivers/staging/android/ion/Kconfig:
```diff
@@ -52,3 +52,39 @@ config ION_OF
+
+config ION_DUMMY_UNMAPED_HEAP
+	bool "Ion dummy driver define a Ion unmaped heap"
+	depends on ION_UNMAPED && ION_DUMMY
```
s/ION_UNMAPED/ION_UNMAP**P**ED/

jockebech commented on drivers/staging/android/ion/Kconfig:
```diff
@@ -52,3 +52,39 @@ config ION_OF
+	bool "Ion dummy driver define a Ion unmaped heap"
+	depends on ION_UNMAPED && ION_DUMMY
+	help
+	  Dummy Ion driver will create a unmaped heap from physical
```
s/Ion/ION/
s/unmaped/unmapped/

jockebech commented on drivers/staging/android/ion/Kconfig:
```diff
@@ -52,3 +52,39 @@ config ION_OF
+	depends on ION_UNMAPED && ION_DUMMY
+	help
+	  Dummy Ion driver will create a unmaped heap from physical
+	  location provided through CONFIG_ION_DUMMY_UNMAPED_BASE and
```
s/CONFIG_ION_DUMMY_UNMAPED_BASE/CONFIG_ION_DUMMY_UNMAP**P**ED_BASE/

jockebech commented on drivers/staging/android/ion/Kconfig:
```diff
@@ -52,3 +52,39 @@ config ION_OF
+	  location provided through CONFIG_ION_DUMMY_UNMAPED_BASE and
+	  CONFIG_ION_DUMMY_UNMAPED_SIZE.
+
+config ION_DUMMY_UNMAPED_BASE
```
s/ION_DUMMY_UNMAPED_BASE/ION_DUMMY_UNMAP**P**ED_BASE/

jockebech commented on drivers/staging/android/ion/Kconfig:
```diff
@@ -52,3 +52,39 @@ config ION_OF
+	  CONFIG_ION_DUMMY_UNMAPED_SIZE.
+
+config ION_DUMMY_UNMAPED_BASE
+	hex "Physical base address of the Ion unmaped heap"
```
s/Ion/ION/
s/unmaped/unmapped/

jockebech commented on drivers/staging/android/ion/Kconfig:
```diff
@@ -52,3 +52,39 @@ config ION_OF
+
+config ION_DUMMY_UNMAPED_BASE
+	hex "Physical base address of the Ion unmaped heap"
+	depends on ION_DUMMY_UNMAPED_HEAP
```
s/ION_DUMMY_UNMAPED_HEAP/ION_DUMMY_UNMAP**P**ED_HEAP/

jockebech commented on drivers/staging/android/ion/Kconfig:
```diff
@@ -52,3 +52,39 @@ config ION_OF
+	depends on ION_DUMMY_UNMAPED_HEAP
+	help
+	  Allows one the statically define an unmaped heap from the
+	  ion dummy driver to exercice unamped heaps buffer managment.
```
s/Ion/ION/
s/unmaped/unmapped/

jockebech commented on drivers/staging/android/ion/Kconfig:
```diff
@@ -52,3 +52,39 @@ config ION_OF
+	depends on ION_DUMMY_UNMAPED_HEAP
+	help
+	  Allows one the statically define an unmaped heap from the
+	  ion dummy driver to exercice unamped heaps buffer managment.
```
s/Ion/ION/
s/unmaped/unmapped/

jockebech commented on drivers/staging/android/ion/Kconfig:
```diff
@@ -52,3 +52,39 @@ config ION_OF
+	  Allows one the statically define an unmaped heap from the
+	  ion dummy driver to exercice unamped heaps buffer managment.
+
+config ION_DUMMY_UNMAPED_SIZE
```
s/ION_DUMMY_UNMAPED_SIZE/ION_DUMMY_UNMAP**P**ED_SIZE/

jockebech commented on drivers/staging/android/ion/Kconfig:
```diff
@@ -52,3 +52,39 @@ config ION_OF
+	  ion dummy driver to exercice unamped heaps buffer managment.
+
+config ION_DUMMY_UNMAPED_SIZE
+	hex "Physical byte size of the Ion unmaped heap"
```
s/Ion/ION/
s/unmaped/unmapped/

jockebech commented on drivers/staging/android/ion/Kconfig:
```diff
@@ -52,3 +52,39 @@ config ION_OF
+
+config ION_DUMMY_UNMAPED_SIZE
+	hex "Physical byte size of the Ion unmaped heap"
+	depends on ION_DUMMY_UNMAPED_HEAP
```
s/ION_DUMMY_UNMAPED_HEAP/ION_DUMMY_UNMAP**P**ED_HEAP/

jockebech commented on drivers/staging/android/ion/Makefile:
```diff
@@ -11,3 +12,4 @@ obj-$(CONFIG_ION_TEGRA) += tegra/
 obj-$(CONFIG_ION_HISI) += hisilicon/
 obj-$(CONFIG_ION_OF) += ion_of.o
 
+obj-$(CONFIG_ION_UNMAPED) += ion_unmaped_heap.o
```
Rename ion_unmaped_heap.o to ion_unmap**p**ed_heap.o

jockebech commented on drivers/staging/android/ion/ion_dummy_driver.c:
```diff
@@ -24,13 +24,18 @@
 #include <linux/io.h>
 #include "ion.h"
 #include "ion_priv.h"
+#include "ion_unmaped_heap.h"
```
Rename ion_unmaped_heap.h -> ion_unmap**p**ed_heap.h

jockebech commented on drivers/staging/android/ion/ion_dummy_driver.c:
```diff
@@ -24,13 +24,18 @@
 static void *carveout_ptr;
 static void *chunk_ptr;
 
+#ifdef CONFIG_ION_DUMMY_UNMAPED_HEAP
```
From hear on to a search and replace from "unmaped" to "unmapped", I won't add more comments about it further down.

etienne-lms replied:
ok :)  thanks.

jockebech commented on drivers/staging/android/ion/ion_dummy_driver.c:
```diff
@@ -154,3 +188,117 @@ static void __exit ion_dummy_exit(void)
+	int rc;
+
+	if (dummy_unmaped_heaps) {
+		// TODO: need to support several nodes ?
```
I don't think that is necessary, not as of now at least. One should be enough for now I think.

etienne-lms replied:
ok, i remove the comment and keep the test.

jockebech commented on drivers/staging/android/ion/ion_dummy_driver.c:
```diff
@@ -154,3 +188,117 @@ static void __exit ion_dummy_exit(void)
+	dummy_unmaped_heaps = devm_kzalloc(&pdev->dev,
+				sizeof(struct ion_heap *) * num_heaps,
+				GFP_KERNEL);
+	if (!dummy_unmaped_pheaps || !dummy_unmaped_heaps) {
```
I'd probably check them one by one, but it's a matter of preference.

jockebech commented on drivers/staging/android/ion/ion_unmaped_heap.c:
```diff
@@ -0,0 +1,345 @@
+#include "ion_priv.h"
+#include "ion_unmaped_heap.h"
+
+/* TODO: could be removed? */
```
If not removed, what about instead using "`ENOMEM`"?

etienne-lms replied:
ok

jockebech commented on drivers/staging/android/ion/ion_unmaped_heap.c:
```diff
@@ -0,0 +1,345 @@
+	struct ion_unmaped_heap *umh =
+		container_of(heap, struct ion_unmaped_heap, heap);
+
+	if (addr == __ALLOCATE_FAIL)
```
Now when I see how this actually is used you cannot probably use ENOMEM as I suggested further up.

jockebech commented on drivers/staging/android/ion/ion_unmaped_heap.c:
```diff
@@ -0,0 +1,345 @@
+	gen_pool_free(umh->pool, addr, size);
+}
+
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(4,9,0))
```
I couldn't see this done in other c-files in Linux. Might be or might not be someone arguing against doing so. But let's keep it for now. Maybe we even can ignore supporting pre 4.9 in the initial version and do that as a separate back-porting activity later on instead?

etienne-lms replied:
i made this coed to be able to port on various linux kernels since ION heap API recently changed (from 4.8 to 4.9). I you feel this code should be strictly dedicated the linux-swg/linux.git "optee" branch, we can remove useless conditioned code.

jockebech commented on drivers/staging/android/ion/ion_unmaped_heap.c:
```diff
@@ -0,0 +1,345 @@
+	struct ion_platform_heap *pheap;
+	enum ion_heap_type type;
+	const char *name;
+	//const char *type_str;
```
Remove?

etienne-lms replied:
yes, removed.

jockebech commented on drivers/staging/android/ion/ion_unmaped_heap.c:
```diff
@@ -0,0 +1,345 @@
+	}
+
+	type = ION_HEAP_TYPE_UNMAPED;
+#if 0
```
For an official merge to linaro-swg/linux you would need to remove this.

etienne-lms replied:
yes, sure.  i remove this.

jockebech commented on drivers/staging/android/ion/ion_unmaped_heap.c:
```diff
@@ -0,0 +1,345 @@
+	}
+#endif
+
+	if (dev)
```
How come there are two ways here? Is it a `dev` sometimes, but no always?

etienne-lms replied:
no, typo, old deprecated code. `dev` is always provided. i fix this.

etienne-lms commented:
As many changes for renaming "unmaped"=>"unmapped", i will squash the commit and propose a new version...

etienne-lms commented on drivers/staging/android/ion/ion_unmapped_heap.c:
```diff
@@ -0,0 +1,261 @@
+	gen_pool_free(umh->pool, addr, size);
+}
+
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(4,9,0))
```
definitely get rid of these kernel version considerations ? (eases backport to 4.8 and earlier kernel releases)

jockebech replied:
Since our master branch is using code newer than 4.9, members seems to use 4.9 LSK OP-TEE and LMG and LHG are moving to 4.9 I'd say that we can remove pre-4.9 code and add that code in backports when we're about to do the backports.

etienne-lms replied:
ok, thanks.

etienne-lms commented on drivers/staging/android/ion/ion_unmapped_heap.c:
```diff
@@ -0,0 +1,261 @@
+	umh->heap.ops = &unmapped_heap_ops;
+	umh->heap.type = ION_HEAP_TYPE_UNMAPPED;
+
+	// TODO: move this out (or the debug)
```
oups.. i forgot to remove this!

jockebech replied:
And it's still here :)

etienne-lms replied:
oups again !

etienne-lms commented:
rebased to latest 'optee' branch.

jockebech commented on drivers/staging/android/ion/ion_unmapped_heap.c:
```diff
@@ -0,0 +1,261 @@
+/*
+ * drivers/staging/android/ion/ion_unmaped_heap.c
```
Does this line belong here?

etienne-lms replied:
it seems to be the ION way.
(i'll fix unmaped => unmapped)

jockebech commented on drivers/staging/android/ion/ion_unmapped_heap.c:
```diff
@@ -0,0 +1,261 @@
+
+	/*
+	 * when user call ION_IOC_ALLOC not with ION_FLAG_CACHED, ion_mmap will
+	 * change prog to pgprot_writecombine itself, so we do not need change
```
prog?

etienne-lms replied:
i will correct the comment. thanks.

jockebech commented on drivers/staging/android/uapi/ion.h:
```diff
@@ -30,6 +30,7 @@ typedef int ion_user_handle_t;
  *				 carveout heap, allocations are physically
  *				 contiguous
  * @ION_HEAP_TYPE_DMA:		 memory allocated via DMA API
+ * @ION_HEAP_TYPE_UNMAPPED:	 memory not mappable into linux address space
```
Is it "not mappable" or "not intended to be mapped"?

etienne-lms replied:
I will replace with "not intended to be mapped".

jockebech commented:
Just some minor cleanup needed (spellings etc), when that is done:
`Reviewed-by: Joakim Bech <joakim.bech@linaro.org>`

etienne-lms commented:
comments addressed, tag applied.

ghost commented:
Hi there, may I know that : 
1. Have this pull been merged into main branch ?
2. Is there any demo dts or config file defined SDP memory start and size ?

jockebech commented:
> Have this pull been merged into main branch ?

No, as long as the PR is open it hasn't been merged, having that said, the intention is to merge it to get it included in the OP-TEE 2.4.0 (April 7th).

> Is there any demo dts or config file defined SDP memory start and size ?

I guess this should be it, @etienne-lms can confirm: https://github.com/OP-TEE/build/pull/112/files

etienne-lms commented:
@ken1979, @jbech-linaro I confirm.
[Current PR](https://github.com/linaro-swg/linux/pull/35) provides SDP memory pool allocation on REE side.
[build PR#112](https://github.com/OP-TEE/build/pull/112/files) enables this SDP support, and sets a test configuration for playing SDP of the QEMU config (for qemu_virt and qemu_armv8). This because is hard codes in the kernel config the SDP memory pool location. In optee_os, a default configuration makes the same area reserved to the SDP memory pool: [df05d4e](https://github.com/OP-TEE/optee_os/commit/df05d4ea7d586234a254d60818154256cac10489).

jforissier commented:
@etienne-lms can you please rebase? (the OP-TEE driver v15 and v16 deltas have been merged since).

etienne-lms commented:
oups...

etienne-lms committed:
staging/ion: add Allwinner ION "secure" heap **not for mainline**

Dumped from:
  https://github.com/loboris/OrangePI-Kernel/tree/master/linux-3.4
  0cc8d855adb457d1860d6e25cb93b6cc75d5a09d
  Author: Sunny <sunny@allwinnertech.com> for Allwinner.

Changes made on original "secure heap" implementation:
- minor coding style: fix includes, empty lines and overlong lines,
  indentation, comment layout.
- Original path modified the ion uapi. We do not attempt to modify
  uapi/ion.h. "secure" (or "domain") heaps are under ID
  ION_HEAP_TYPE_CUSTOM + 1 (legacy 'secure heap type' value).

Signed-off-by: Etienne Carriere <etienne.carriere@linaro.org>
Reviewed-by: Joakim Bech <joakim.bech@linaro.org>

etienne-lms committed:
ion: "unmapped" heap for secure data path **not for mainline**

OP-TEE/SDP (Secure Data Path) memory pools are created through ION
secure type heap" from Allwinner. This change renames "secure" into
"unmapped" as, from Linux point of view, the heap constraint is
manipulating unmapped memory pools/buffers.

"Unmapped" heap support is integrated in ION UAPI (actually this was
the Allwinner initial proposal) and ION DT parsing support.

Based in work from Sunny <sunny@allwinnertech.com> for Allwinner.

Signed-off-by: Etienne Carriere <etienne.carriere@linaro.org>
Reviewed-by: Joakim Bech <joakim.bech@linaro.org>

etienne-lms committed:
ion: unmapped heap support in ion dummy driver **not for mainline**

Add configuration ION_DUMMY_UNMAPPED_HEAP to enable optional definition
of a statically defined "unmapped" heap for test purpose: kernel config
must provide the memory pool base address and byte size.

Signed-off-by: Etienne Carriere <etienne.carriere@linaro.org>
Reviewed-by: Joakim Bech <joakim.bech@linaro.org>

etienne-lms commented:
OP-TEE/build#112 also rebased.
